### ISSUE TITLE
Autologin with nextcloud credentials

### ISSUE THREAD

**Issue by diroots** [pierre-alain-b/rainloop-nextcloud#221]
We are using nextcloud credentials for rainloop to autologin. with the option "Attempt to automatically login users with their Nextcloud username and password, or user-defined credentials, if set."

we also use nextcloud's default cookie / session lifetime from there : https://github.com/nextcloud/server/blob/master/config/config.sample.php#L247

so users do not need to re-log in to nextcloud each time they return to their nextcloud account. 

If a user log in to nextcloud then switch to rainloop, autologin works well
if a user returns to his nextcloud account, and then go back to the rainloop tab, the autologin fails with the following logs (sometime only first log block) : 

```
--- FlushLogCache: WriteOnErrorOnly
[11:32:16.217][65a2a80f] INFO[DATA]: [DATE:23.02.21][OFFSET:-00][RL:1.14.0][PHP:7.4.15][IP:xxx][PID:1527][nginx/1.19.7][fpm-fcgi]
[11:32:16.217][65a2a80f] INFO[DATA]: [Suhosin:off][APC:off][MB:on][PDO:sqlite,mysql,pgsql][cloud:true][Streams:tcp,udp,unix,udg,ssl,tls,tlsv1.0,tlsv1.1,tlsv1.2,tlsv1.3]
[11:32:16.217][65a2a80f] REQUEST[NOTE]: [GET] https://xxx/apps/rainloop/app/?OwnCloudAuth
--- FlushLogCache: Trigger
[11:32:17.235][65a2a80f] INFO[NOTICE]: RainLoop\Exceptions\ClientException: InvalidInputArgument[903] in /var/www/html/custom_apps/rainloop/app/rainloop/v/1.14.0/app/libraries/RainLoop/Actions.php:2308
Stack trace:
#0 /var/www/html/custom_apps/rainloop/app/rainloop/v/1.14.0/app/libraries/RainLoop/ServiceActions.php(277): RainLoop\Actions->LoginProcess('xxx@xxx...', '')
#1 /var/www/html/custom_apps/rainloop/app/rainloop/v/1.14.0/app/libraries/RainLoop/Service.php(150): RainLoop\ServiceActions->ServiceOwnCloudAuth('')
#2 /var/www/html/custom_apps/rainloop/app/rainloop/v/1.14.0/app/libraries/RainLoop/Service.php(62): RainLoop\Service->localHandle()
#3 /var/www/html/custom_apps/rainloop/app/rainloop/v/1.14.0/app/libraries/RainLoop/Service.php(85): RainLoop\Service->__construct()
#4 /var/www/html/custom_apps/rainloop/app/rainloop/v/1.14.0/app/handle.php(94): RainLoop\Service::Handle()
#5 /var/www/html/custom_apps/rainloop/app/rainloop/v/1.14.0/include.php(228): include('/var/www/html/c...')
#6 /var/www/html/custom_apps/rainloop/app/index.php(23): include('/var/www/html/c...')
#7 /var/www/html/custom_apps/rainloop/lib/Controller/PageController.php(105): include('/var/www/html/c...')
#8 /var/www/html/custom_apps/rainloop/lib/Controller/PageController.php(59): OCA\RainLoop\Controller\PageController->app()
#9 /var/www/html/lib/private/AppFramework/Http/Dispatcher.php(170): OCA\RainLoop\Controller\PageController->appGet()
#10 /var/www/html/lib/private/AppFramework/Http/Dispatcher.php(100): OC\AppFramework\Http\Dispatcher->executeController(Object(OCA\RainLoop\Controller\PageController), 'appGet')
#11 /var/www/html/lib/private/AppFramework/App.php(137): OC\AppFramework\Http\Dispatcher->dispatch(Object(OCA\RainLoop\Controller\PageController), 'appGet')
#12 /var/www/html/lib/private/AppFramework/Routing/RouteActionHandler.php(47): OC\AppFramework\App::main('PageController', 'appGet', Object(OC\AppFramework\DependencyInjection\DIContainer), Array)
#13 [internal function]: OC\AppFramework\Routing\RouteActionHandler->__invoke(Array)
#14 /var/www/html/lib/private/Route/Router.php(297): call_user_func(Object(OC\AppFramework\Routing\RouteActionHandler), Array)
#15 /var/www/html/lib/base.php(1010): OC\Route\Router->match('/apps/rainloop/...')
#16 /var/www/html/index.php(37): OC::handleRequest()
#17 {main}
[11:32:17.238][65a2a80f] INFO[DATA]: Location: ./
[11:32:17.240][65a2a80f] INFO[MEMORY]: Memory peak usage: 2MB
[11:32:17.240][65a2a80f] INFO[TIME]: Time delta: 1.0349881649017

--- FlushLogCache: WriteOnErrorOnly
[11:32:20.067][257de221] INFO[DATA]: [DATE:23.02.21][OFFSET:-00][RL:1.14.0][PHP:7.4.15][IP:xxx][PID:3155][nginx/1.19.7][fpm-fcgi]
[11:32:20.067][257de221] INFO[DATA]: [Suhosin:off][APC:off][MB:on][PDO:sqlite,mysql,pgsql][cloud:true][Streams:tcp,udp,unix,udg,ssl,tls,tlsv1.0,tlsv1.1,tlsv1.2,tlsv1.3]
[11:32:20.067][257de221] REQUEST[NOTE]: [POST] https://xxx/apps/rainloop/app/?/Ajax/&q[]=/0/
[11:32:20.067][257de221] AJAX[NOTE]: Action: DoFolders
[11:32:20.067][257de221] POST[DATA]: {"Action":"Folders","XToken":"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"}
--- FlushLogCache: Trigger
[11:32:20.068][257de221] INFO[NOTICE]: RainLoop\Exceptions\ClientException: AuthError[102] in /var/www/html/custom_apps/rainloop/app/rainloop/v/1.14.0/app/libraries/RainLoop/Actions.php:1376
Stack trace:
#0 /var/www/html/custom_apps/rainloop/app/rainloop/v/1.14.0/app/libraries/RainLoop/Actions.php(1429): RainLoop\Actions->GetAccountFromCustomToken('', true, true, true)
#1 /var/www/html/custom_apps/rainloop/app/rainloop/v/1.14.0/app/libraries/RainLoop/Actions.php(9057): RainLoop\Actions->getAccountFromToken()
#2 /var/www/html/custom_apps/rainloop/app/rainloop/v/1.14.0/app/libraries/RainLoop/Actions.php(5368): RainLoop\Actions->initMailClientConnection()
#3 /var/www/html/custom_apps/rainloop/app/rainloop/v/1.14.0/app/libraries/RainLoop/ServiceActions.php(174): RainLoop\Actions->DoFolders()
#4 /var/www/html/custom_apps/rainloop/app/rainloop/v/1.14.0/app/libraries/RainLoop/Service.php(150): RainLoop\ServiceActions->ServiceAjax('')
#5 /var/www/html/custom_apps/rainloop/app/rainloop/v/1.14.0/app/libraries/RainLoop/Service.php(62): RainLoop\Service->localHandle()
#6 /var/www/html/custom_apps/rainloop/app/rainloop/v/1.14.0/app/libraries/RainLoop/Service.php(85): RainLoop\Service->__construct()
#7 /var/www/html/custom_apps/rainloop/app/rainloop/v/1.14.0/app/handle.php(94): RainLoop\Service::Handle()
#8 /var/www/html/custom_apps/rainloop/app/rainloop/v/1.14.0/include.php(228): include('/var/www/html/c...')
#9 /var/www/html/custom_apps/rainloop/app/index.php(23): include('/var/www/html/c...')
#10 /var/www/html/custom_apps/rainloop/lib/Controller/PageController.php(105): include('/var/www/html/c...')
#11 /var/www/html/custom_apps/rainloop/lib/Controller/PageController.php(67): OCA\RainLoop\Controller\PageController->app()
#12 /var/www/html/lib/private/AppFramework/Http/Dispatcher.php(170): OCA\RainLoop\Controller\PageController->appPost()
#13 /var/www/html/lib/private/AppFramework/Http/Dispatcher.php(100): OC\AppFramework\Http\Dispatcher->executeController(Object(OCA\RainLoop\Controller\PageController), 'appPost')
#14 /var/www/html/lib/private/AppFramework/App.php(137): OC\AppFramework\Http\Dispatcher->dispatch(Object(OCA\RainLoop\Controller\PageController), 'appPost')
#15 /var/www/html/lib/private/AppFramework/Routing/RouteActionHandler.php(47): OC\AppFramework\App::main('PageController', 'appPost', Object(OC\AppFramework\DependencyInjection\DIContainer), Array)
#16 [internal function]: OC\AppFramework\Routing\RouteActionHandler->__invoke(Array)
#17 /var/www/html/lib/private/Route/Router.php(297): call_user_func(Object(OC\AppFramework\Routing\RouteActionHandler), Array)
#18 /var/www/html/lib/base.php(1010): OC\Route\Router->match('/apps/rainloop/...')
#19 /var/www/html/index.php(37): OC::handleRequest()
#20 {main}
[11:32:20.068][257de221] AJAX[DATA]: {"Action":"Folders","Result":false,"ErrorCode":102,"ErrorMessage":"","ErrorMessageAdditional":"","ClearAuth":true,"Time":13}
[11:32:20.083][257de221] INFO[MEMORY]: Memory peak usage: 6MB
[11:32:20.083][257de221] INFO[TIME]: Time delta: 0.02790093421936
```

we did try to align the rainloop's setting "autologout" to 1440 minuts in application.ini (it was initially set to 30, so i assumed it was minuts) but still this do not work anymore. i didn't find any other rainloop settings to keep session available as long as NC's session

it used to work in older rainloop plugin version, but since latest 7.0.3 on latest nc 19.0.8 it doesn't autologin after some time. 

i believe this is linked to the new change to store credentials in session only,  right? 

any idea howto get the feature back ? as if not, the "Attempt to automatically login users with their Nextcloud username and password, or user-defined credentials, if set." option can be quite useless,...

**Comment by ogmkp** [comment]
Observed since NC 18 (on my server).

**Comment by diroots** [comment]
@pierre-alain-b any idea on howto bypass this problem? 

could this be because a logout action is fired, emptying the rainloop-autologin-password session value here : https://github.com/pierre-alain-b/rainloop-nextcloud/blob/master/lib/Util/RainLoopHelper.php#L235

this seems to happen even if the autologout settings value is set to 1440 minutes (so that would normally prevent logout for one day,... ) is the logout function fired whenever a browser tab is unloaded or something like this?

**Comment by pierre-alain-b** [comment]
Could you try with the new version which embarks Rainloop 1.15 to see if the issue is still there?

**Comment by ogmkp** [comment]
It's ok for me, thank you.

**Comment by diroots** [comment]
sorry, but with latest version of app embedding the latest rainloop, i still get the following : 

```
[15:31:37.150][b6a12dd9] INFO[DATA]: [DATE:09.03.21][OFFSET:-00][RL:1.15.0][PHP:7.4.15][IP:xxxipv6][PID:68][nginx/1.19.7][fpm-fcgi]
[15:31:37.150][b6a12dd9] INFO[DATA]: [APC:off][MB:on][PDO:sqlite,mysql,pgsql][cloud:true][Streams:tcp,udp,unix,udg,ssl,tls,tlsv1.0,tlsv1.1,tlsv1.2,tlsv1.3]
[15:31:37.150][b6a12dd9] REQUEST[NOTE]: [GET] https://domain.tld/apps/rainloop/app/?OwnCloudAuth
[15:31:37.171][b6a12dd9] INFO[NOTICE]: RainLoop\Exceptions\ClientException: InvalidInputArgument[903] in /var/www/html/custom_apps/rainloop/app/rainloop/v/1.15.0/app/libraries/RainLoop/Actions.php:2307
Stack trace:
#0 /var/www/html/custom_apps/rainloop/app/rainloop/v/1.15.0/app/libraries/RainLoop/ServiceActions.php(277): RainLoop\Actions->LoginProcess('xxx@xxx...', '')
#1 /var/www/html/custom_apps/rainloop/app/rainloop/v/1.15.0/app/libraries/RainLoop/Service.php(150): RainLoop\ServiceActions->ServiceOwnCloudAuth('')
#2 /var/www/html/custom_apps/rainloop/app/rainloop/v/1.15.0/app/libraries/RainLoop/Service.php(62): RainLoop\Service->localHandle()
#3 /var/www/html/custom_apps/rainloop/app/rainloop/v/1.15.0/app/libraries/RainLoop/Service.php(85): RainLoop\Service->__construct()
#4 /var/www/html/custom_apps/rainloop/app/rainloop/v/1.15.0/app/handle.php(94): RainLoop\Service::Handle()
#5 /var/www/html/custom_apps/rainloop/app/rainloop/v/1.15.0/include.php(243): include('/var/www/html/c...')
#6 /var/www/html/custom_apps/rainloop/app/index.php(23): include('/var/www/html/c...')
#7 /var/www/html/custom_apps/rainloop/lib/Controller/PageController.php(105): include('/var/www/html/c...')
#8 /var/www/html/custom_apps/rainloop/lib/Controller/PageController.php(59): OCA\RainLoop\Controller\PageController->app()
#9 /var/www/html/lib/private/AppFramework/Http/Dispatcher.php(170): OCA\RainLoop\Controller\PageController->appGet()
#10 /var/www/html/lib/private/AppFramework/Http/Dispatcher.php(100): OC\AppFramework\Http\Dispatcher->executeController(Object(OCA\RainLoop\Controller\PageController), 'appGet')
#11 /var/www/html/lib/private/AppFramework/App.php(137): OC\AppFramework\Http\Dispatcher->dispatch(Object(OCA\RainLoop\Controller\PageController), 'appGet')
#12 /var/www/html/lib/private/AppFramework/Routing/RouteActionHandler.php(47): OC\AppFramework\App::main('PageController', 'appGet', Object(OC\AppFramework\DependencyInjection\DIContainer), Array)
#13 [internal function]: OC\AppFramework\Routing\RouteActionHandler->__invoke(Array)
#14 /var/www/html/lib/private/Route/Router.php(297): call_user_func(Object(OC\AppFramework\Routing\RouteActionHandler), Array)
#15 /var/www/html/lib/base.php(1010): OC\Route\Router->match('/apps/rainloop/...')
#16 /var/www/html/index.php(37): OC::handleRequest()
#17 {main}
[15:31:37.171][b6a12dd9] INFO[DATA]: Location: ./
[15:31:37.236][b6a12dd9] INFO[MEMORY]: Memory peak usage: 6MB
[15:31:37.236][b6a12dd9] INFO[TIME]: Time delta: 0.30154705047607
```

and if a user was connected yesterday in nextcloud, then came back today on his nc account, it is still connected, but switching to rainloop, credentials are still required. so autologin with nc credentials do not work with default setup of nc / rainloop.

idea : could the rainloop_autologin_password be stored in the remember_login_cookie of NC so this maybe won't happen?


@pierre-alain-b can the issue be reopened as it is not fixed for me, the reporter of this issue?

**Comment by loxK** [comment]
@diroots it does work. Even on Nextcloud 21. But it is also true that if you have logged in Nextcloud since a while, it won't work then. It is needed to log out and back in Nextcloud to see it working again.
This is the behavior since it has been released, and I also find it annoying. 

But it is another issue that I think should be reported and properly explained separately.

@pierre-alain-b any thoughts about this buggy behavior ?

**Comment by nkinkade** [comment]
I'm seeing this as well, and have debugged enough to verify that the session variable `rainloop-autologin-password` is disappearing. I have also verified that this is not because the [logout hook](https://github.com/pierre-alain-b/rainloop-nextcloud/blob/master/lib/Util/RainLoopHelper.php#L235) is somehow being called, setting `rainloop-autologin-password` to an empty string. I did come across this:

https://github.com/nextcloud/server/issues/2081#issuecomment-322240642

This comment from @MorrisJobke would make it appear that session data is completely ephemeral and can never be relied upon existing at any given moment, and that Nextcloud might delete an session and create a new one at any given time, removing all prior session data. This doesn't make complete sense to me. If my login session to Nextcloud has not expired, then I would not expect for session variables set by an app (rainloop in this case) to disappear if my Nextcloud login session has also not expired.

I may be missing something obvious here, but my expectation would be that as long as my Nextcloud login session has not expired, that session variables set by an app would also persist for as long as that login session does. But it appear that what is actually happening is that I remain logged into Nextcloud but invisibly in the background Nextcloud is deleting a session and creating a new one.
